core/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java

```diff
@@ -150,24 +155,34 @@ public AllocateUnassignedDecision makeAllocationDecision(final ShardRouting unas
         final RoutingNodes routingNodes = allocation.routingNodes();
         final boolean explain = allocation.debugDecision();
         // pre-check if it can be allocated to any node that currently exists, so we won't list the store for it for nothing
-        Tuple<Decision, Map<String, Decision>> allocateDecision = canBeAllocatedToAtLeastOneNode(unassignedShard, allocation, explain);
-        if (allocateDecision.v1().type() != Decision.Type.YES) {
+        Tuple<Decision, Map<String, NodeAllocationResult>> result = canBeAllocatedToAtLeastOneNode(unassignedShard, allocation);
+        Decision allocateDecision = result.v1();
+        if (allocateDecision.type() != Decision.Type.YES) {
+            // only return early if we are not in explain mode, because if we are in explain mode,
```

Comment:
The issue is with fetching.  Suppose `canBeAllocatedToAtleastOneNode` returns a non-YES decision.  If this is the case from the start of the ES instance, then we will have never before fetched shard data (b/c `canBeAllocatedToAtleastOneNode`) aborted us from doing so.  Now suppose we execute an explain request.  If we didn't abort early here, then the explain call would've triggered a shard fetch, and the response we would get from the explain API is "pending shard fetch".  

I don't believe this is accurate, as the explain API is not reflecting reality (i.e. its telling us its pending shard fetching when that's not the reality of allocation, the reality of allocation is that it never bothered trying to fetch shard data due to a NO or THROTTLE decision from `canBeAllocatedToAtLeastOneNode`).  I don't think the explain API should be the cause of a shard fetch for the `PrimaryShardAllocator` nor the `ReplicaShardAllocator`.  

That was the reasoning here for aborting early, even in explain mode.  One way to avoid the explain API triggering a shard fetch and still return node decisions even if `canBeAllocatedToAtleastOneNode` returns NO is to add a method to the `ReplicaShardsAllocator` to check if a shard data fetch has ever been triggered.  I tried this strategy in this commit: aa31f4c72bd4f1b0febab835b73542c00364db32.  Let me know what you think.